src/prefect/new_task_engine.py

```diff
@@ -546,32 +549,45 @@ def run_task_sync(
     # This is a context manager that keeps track of the run of the task run.
     with engine.start(task_run_id=task_run_id, dependencies=dependencies) as run:
         with run.enter_run_context():
-            with transaction(record=Record()) as txn:
-                txn.add_task(run.task, run.task_run.id)
-                run.begin_run()
-                while run.is_running():
-                    # enter run context on each loop iteration to ensure the context
-                    # contains the latest task run metadata
-                    with run.enter_run_context():
-                        try:
-                            # This is where the task is actually run.
-                            with timeout(seconds=run.task.timeout_seconds):
-                                call_args, call_kwargs = parameters_to_args_kwargs(
-                                    task.fn, run.parameters or {}
-                                )
-                                run.logger.debug(
-                                    f"Executing task {task.name!r} for task run {run.task_run.name!r}..."
-                                )
-                                result = cast(R, task.fn(*call_args, **call_kwargs))  # type: ignore
-
-                            # If the task run is successful, finalize it.
-                            run.handle_success(result)
-                        except RollBack as exc:
-                            run.handle_rollback(exc, transaction=txn)
-                        except TimeoutError as exc:
-                            run.handle_timeout(exc)
-                        except Exception as exc:
-                            run.handle_exception(exc)
+            run.begin_run()
+            while run.is_running():
+                # enter run context on each loop iteration to ensure the context
+                # contains the latest task run metadata
+                with run.enter_run_context():
+                    try:
+                        # This is where the task is actually run.
+                        with timeout(seconds=run.task.timeout_seconds):
+                            call_args, call_kwargs = parameters_to_args_kwargs(
+                                task.fn, run.parameters or {}
+                            )
+                            run.logger.debug(
+                                f"Executing task {task.name!r} for task run {run.task_run.name!r}..."
+                            )
+                            result_factory = getattr(
+                                TaskRunContext.get(), "result_factory", None
+                            )
+                            with transaction(
```

Comment:
The only thing that really changed in this big block is that I moved where I entered and exited the transaction, causing a large indention change